Allow getaddrinfo to resolve localhost or local hostname

jukkar opened the pull request:
Allow user to query localhost name and get a localhost address 127.0.0.1 or ::1 back. This is only enabled if loopback driver is used as the use case does not make much sense otherwise.
    
Make sure that if user has enabled hostname support, we are able to query it and get a local IP address back.

Fixes #93822

jukkar committed:
net: dns: Resolve local hostname

Make sure that if user has enabled hostname support, we are
able to query it and get a local IP address back.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Resolve localhost name

Allow user to query localhost name and get a localhost address
127.0.0.1 or ::1 back. This is only enabled if loopback driver
is used as the use case does not make much sense otherwise.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: dns: Add tests for resolving hostname or localhost

Make sure we are able to resolve the localhost or local hostname.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -1827,6 +1827,107 @@ int dns_resolve_name_internal(struct dns_resolve_context *ctx,
+		/* If the hostname is the same as the query, then
+		 * return a local address.
+		 */
+		if (strncmp(hostname, query, strlen(hostname)) == 0) {
```
Calling `strlen` to limit the length in `strncmp` isn't useful.

pdgendt commented on tests/net/lib/dns_resolve/src/main.c:
Indent using tabs.

jukkar committed:
net: sockets: Allow test to override DNS timeout

The minimum 1 sec timeout for getaddrinfo() call is too long
when executing queries from network tests. So allow test to override
the minimum and set it to very low value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Updated according to comments

pdgendt commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -1827,6 +1827,107 @@ int dns_resolve_name_internal(struct dns_resolve_context *ctx,
 
+	/* If we get a query to localhost, then short circuit early */
+	if ((IS_ENABLED(CONFIG_NET_LOOPBACK) || IS_ENABLED(CONFIG_NET_TEST)) &&
+	    memcmp(query, "localhost", sizeof("localhost")) == 0) {
```
Isn't this already an issue if the length of `query` would be smaller than the length of `"localhost"`?

rlubos replied:
Is the assumption that `query` may not be a null-terminated string valid though? We use `strlen(query)` just above.
